File: python/pinarkive_client.py
import requests
from typing import Optional, Dict, Any, List

class PinarkiveClient:
    def __init__(self, token: Optional[str] = None, api_key: Optional[str] = None, base_url: str = 'https://api.pinarkive.com/api/v2'):
        self.base_url = base_url
        self.token = token
        self.api_key = api_key
        self.session = requests.Session()

    def _headers(self) -> Dict[str, str]:
        headers = {}
        if self.token:
            headers['Authorization'] = f'Bearer {self.token}'
        elif self.api_key:
            headers['X-API-Key'] = self.api_key
        return headers

    # --- Authentication ---
    def login(self, email: str, password: str) -> Any:
        return self.session.post(
            f'{self.base_url}/auth/login',
            json={'email': email, 'password': password}
        )

    def signup(self, data: Dict[str, Any], locale: Optional[str] = None, refCode: Optional[str] = None) -> Any:
        params: Dict[str, str] = {}
        if locale:
            params['locale'] = locale
        if refCode:
            params['refCode'] = refCode
        return self.session.post(
            f'{self.base_url}/auth/signup',
            json=data,
            params=params
        )

    def logout(self) -> Any:
        return self.session.post(
            f'{self.base_url}/auth/logout',
            headers=self._headers()
        )

    # --- File Management ---
    def upload_file(self, file_path: str) -> Any:
        with open(file_path, 'rb') as f:
            files = {'file': f}
            return self.session.post(
                f'{self.base_url}/files',
                files=files,
                headers=self._headers()
            )

    def upload_directory(self, dir_path: str) -> Any:
        return self.session.post(
            f'{self.base_url}/files/directory',
            json={'dirPath': dir_path},
            headers=self._headers()
        )

    def pin_cid(self, cid: str) -> Any:
        return self.session.post(
            f'{self.base_url}/files/pin/{cid}',
            headers=self._headers()
        )

    def remove_file(self, cid: str) -> Any:
        return self.session.delete(
            f'{self.base_url}/files/remove/{cid}',
            headers=self._headers()
        )

    # --- User Profile ---
    def get_profile(self) -> Any:
        return self.session.get(
            f'{self.base_url}/users/me',
            headers=self._headers()
        )

    def update_profile(self, data: Dict[str, Any]) -> Any:
        return self.session.put(
            f'{self.base_url}/users/me',
            json=data,
            headers=self._headers()
        )

    def list_uploads(self, page: int = 1, limit: int = 10) -> Any:
        return self.session.get(
            f'{self.base_url}/users/me/uploads',
            params={'page': page, 'limit': limit},
            headers=self._headers()
        )

    def delete_upload(self, cid: str) -> Any:
        return self.session.delete(
            f'{self.base_url}/users/me/uploads/{cid}',
            headers=self._headers()
        )

    def get_referrals(self) -> Any:
        return self.session.get(
            f'{self.base_url}/users/me/referrals',
            headers=self._headers()
        )

    # --- Token Management ---
    def generate_token(self, name: str, permissions: Optional[List[str]] = None) -> Any:
        data: Dict[str, Any] = {'name': name}
        if permissions:
            data['permissions'] = permissions
        return self.session.post(
            f'{self.base_url}/tokens/generate',
            json=data,
            headers=self._headers()
        )

    def list_tokens(self) -> Any:
        return self.session.get(
            f'{self.base_url}/tokens/list',
            headers=self._headers()
        )

    def revoke_token(self, name: str) -> Any:
        return self.session.delete(
            f'{self.base_url}/tokens/revoke/{name}',
            headers=self._headers()
        )

    # --- Status and Monitoring ---
    def get_status(self, cid: str) -> Any:
        return self.session.get(
            f'{self.base_url}/status/{cid}',
            headers=self._headers()
        )

    def get_allocations(self, cid: str) -> Any:
        return self.session.get(
            f'{self.base_url}/status/allocations/{cid}',
            headers=self._headers()
        )

File: go/pinarkive_client.go
package pinarkive

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
)

type PinarkiveClient struct {
	BaseURL string
	Token   string
	APIKey  string
	Client  *http.Client
}

func NewPinarkiveClient(token, apiKey, baseURL string) *PinarkiveClient {
	if baseURL == "" {
		baseURL = "https://api.pinarkive.com/api/v2"
	}
	return &PinarkiveClient{
		BaseURL: baseURL,
		Token:   token,
		APIKey:  apiKey,
		Client:  &http.Client{},
	}
}

func (c *PinarkiveClient) headers() http.Header {
	headers := http.Header{}
	if c.Token != "" {
		headers.Set("Authorization", "Bearer "+c.Token)
	} else if c.APIKey != "" {
		headers.Set("X-API-Key", c.APIKey)
	}
	return headers
}

// --- Authentication ---
func (c *PinarkiveClient) Login(email, password string) (*http.Response, error) {
	data := map[string]string{"email": email, "password": password}
	return c.postJson("/auth/login", data, nil)
}

func (c *PinarkiveClient) Signup(data map[string]interface{}, locale, refCode string) (*http.Response, error) {
	params := ""
	if locale != "" || refCode != "" {
		params = "?"
		if locale != "" {
			params += "locale=" + locale
		}
		if refCode != "" {
			if locale != "" {
				params += "&"
			}
			params += "refCode=" + refCode
		}
	}
	return c.postJson("/auth/signup"+params, data, nil)
}

func (c *PinarkiveClient) Logout() (*http.Response, error) {
	return c.postJson("/auth/logout", nil, c.headers())
}

// --- File Management ---
func (c *PinarkiveClient) UploadFile(filePath string) (*http.Response, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var b bytes.Buffer
	w := multipart.NewWriter(&b)
	fw, err := w.CreateFormFile("file", file.Name())
	if err != nil {
		return nil, err
	}
	if _, err = io.Copy(fw, file); err != nil {
		return nil, err
	}
	w.Close()
	url := c.BaseURL + "/files"
	req, err := http.NewRequest("POST", url, &b)
	if err != nil {
		return nil, err
	}
	req.Header = c.headers()
	req.Header.Set("Content-Type", w.FormDataContentType())
	return c.Client.Do(req)
}

func (c *PinarkiveClient) UploadDirectory(dirPath string) (*http.Response, error) {
	data := map[string]string{"dirPath": dirPath}
	return c.postJson("/files/directory", data, c.headers())
}

func (c *PinarkiveClient) PinCid(cid string) (*http.Response, error) {
	return c.postJson("/files/pin/"+cid, nil, c.headers())
}

func (c *PinarkiveClient) RemoveFile(cid string) (*http.Response, error) {
	url := c.BaseURL + "/files/remove/" + cid
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header = c.headers()
	return c.Client.Do(req)
}

// --- User Profile ---
func (c *PinarkiveClient) GetProfile() (*http.Response, error) {
	return c.get("/users/me", c.headers())
}

func (c *PinarkiveClient) UpdateProfile(data map[string]interface{}) (*http.Response, error) {
	return c.putJson("/users/me", data, c.headers())
}

func (c *PinarkiveClient) ListUploads(page, limit int) (*http.Response, error) {
	url := fmt.Sprintf("/users/me/uploads?page=%d&limit=%d", page, limit)
	return c.get(url, c.headers())
}

func (c *PinarkiveClient) DeleteUpload(cid string) (*http.Response, error) {
	url := "/users/me/uploads/" + cid
	return c.delete(url, c.headers())
}

func (c *PinarkiveClient) GetReferrals() (*http.Response, error) {
	return c.get("/users/me/referrals", c.headers())
}

// --- Token Management ---
func (c *PinarkiveClient) GenerateToken(label, name string, expiresInDays int) (*http.Response, error) {
	data := map[string]interface{}{"label": label, "name": name}
	if expiresInDays > 0 {
		data["expiresInDays"] = expiresInDays
	}
	return c.postJson("/tokens/generate", data, c.headers())
}

func (c *PinarkiveClient) ListTokens() (*http.Response, error) {
	return c.get("/tokens/list", c.headers())
}

func (c *PinarkiveClient) RevokeToken(name string) (*http.Response, error) {
	url := "/tokens/revoke/" + name
	return c.delete(url, c.headers())
}

// --- Status and Monitoring ---
func (c *PinarkiveClient) GetStatus(cid string) (*http.Response, error) {
	url := "/status/" + cid
	return c.get(url, c.headers())
}

func (c *PinarkiveClient) GetAllocations(cid string) (*http.Response, error) {
	url := "/status/allocations/" + cid
	return c.get(url, c.headers())
}

// --- Helper HTTP methods ---
func (c *PinarkiveClient) get(path string, headers http.Header) (*http.Response, error) {
	url := c.BaseURL + path
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header = headers
	return c.Client.Do(req)
}

func (c *PinarkiveClient) postJson(path string, data interface{}, headers http.Header) (*http.Response, error) {
	url := c.BaseURL + path
	var body io.Reader
	if data != nil {
		b, _ := json.Marshal(data)
		body = bytes.NewBuffer(b)
	}
	req, err := http.NewRequest("POST", url, body)
	if err != nil {
		return nil, err
	}
	req.Header = headers
	if data != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	return c.Client.Do(req)
}

func (c *PinarkiveClient) putJson(path string, data interface{}, headers http.Header) (*http.Response, error) {
	url := c.BaseURL + path
	b, _ := json.Marshal(data)
	req, err := http.NewRequest("PUT", url, bytes.NewBuffer(b))
	if err != nil {
		return nil, err
	}
	req.Header = headers
	req.Header.Set("Content-Type", "application/json")
	return c.Client.Do(req)
}

func (c *PinarkiveClient) delete(path string, headers http.Header) (*http.Response, error) {
	url := c.BaseURL + path
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header = headers
	return c.Client.Do(req)
}